File: nodejs/src/utils/math.js
// utils/math.js - Deterministic, safe math evaluator.
// Parses arithmetic ( + - * / ^ % parentheses, decimals, scientific notation )
// WITHOUT using eval(), so it can never execute arbitrary code. Returns null
// when the input is not a clean math expression, so callers fall back to the
// model.

// Extract a bare arithmetic expression from a line of chat text.
export function extractMathExpression(text) {
  const s = String(text || '').trim();
  // Strip a leading question/math verb phrase if present.
  const cleaned = s.replace(
    /^(?:what(?:'s| is| are)|what's|compute|calculate|solve|evaluate|equals?)\b\s*/i,
    ''
  ).replace(/\??$/, '').trim();
  const expr = cleaned.replace(/[?,!.]$/, '').trim();
  if (!expr) return null;
  // Guard: must be purely numeric + arithmetic operators + parens + spaces.
  if (!/^[0-9+\-*\/^%().\s]+$/.test(expr)) return null;
  if (!/\d/.test(expr)) return null;
  if (/[a-zA-Z]/.test(expr)) return null;
  // Must contain at least one operator (otherwise "2" alone isn't a math ask).
  if (!/[\+\-\*\/\^%]/.test(expr)) return null;
  return expr;
}

// Tokenize an expression into numbers and operators.
function tokenize(expr) {
  const tokens = [];
  let i = 0;
  const n = expr.length;
  while (i < n) {
    const c = expr[i];
    if (/\s/.test(c)) { i++; continue; }
    if ('+-*/^%()'.includes(c)) {
      tokens.push(c);
      i++;
      continue;
    }
    // read a number (int, decimal, scientific)
    const num = expr.slice(i).match(/^(\d+(?:\.\d+)?|\.\d+)([eE][+-]?\d+)?/);
    if (!num) return null;
    tokens.push(parseFloat(num[0]));
    i += num[0].length;
  }
  return tokens;
}

// Recursive-descent parser with correct precedence:
//   expr   := term (('+'|'-') term)*
//   term   := power (('*'|'/'|'%') power)*
//   power  := unary ('^' power)?        right-associative
//   unary  := ('+'|'-') unary | factor
//   factor := '(' expr ')' | number
class Parser {
  constructor(tokens) {
    this.tokens = tokens;
    this.pos = 0;
  }
  peek() { return this.tokens[this.pos]; }
  next() { return this.tokens[this.pos++]; }
  expect(val) {
    const t = this.next();
    if (t !== val) throw new Error('syntax error');
  }
  parseExpr() { return this.parseTerm(); }
  parseTerm() {
    let left = this.parseMul();
    while (['+', '-'].includes(this.peek())) {
      const op = this.next();
      const right = this.parseMul();
      left = op === '+' ? left + right : left - right;
    }
    return left;
  }
  parseMul() {
    let left = this.parsePower();
    while (['*', '/', '%'].includes(this.peek())) {
      const op = this.next();
      const right = this.parsePower();
      if (op === '*') left = left * right;
      else if (op === '/') left = left / right;
      else left = left % right;
    }
    return left;
  }
  parsePower() {
    let left = this.parseUnary();
    if (this.peek() === '^') {
      this.next();
      const right = this.parsePower();
      left = Math.pow(left, right);
    }
    return left;
  }
  parseUnary() {
    if (this.peek() === '+') { this.next(); return this.parseUnary(); }
    if (this.peek() === '-') {
      this.next();
      return -this.parseUnary();
    }
    return this.parseFactor();
  }
  parseFactor() {
    const t = this.peek();
    if (t === '(') {
      this.next();
      const v = this.parseExpr();
      this.expect(')');
      return v;
    }
    if (typeof t === 'number') {
      this.next();
      return t;
    }
    throw new Error('syntax error');
  }
}

export function evaluateMath(expr) {
  try {
    const tokens = tokenize(expr);
    if (!tokens || !tokens.length) return null;
    const p = new Parser(tokens);
    const result = p.parseExpr();
    if (p.pos !== tokens.length) return null; // leftover tokens = not a clean expr
    if (!Number.isFinite(result)) return null;
    // Round to avoid floating point noise like 0.30000000000000004
    const rounded = Math.round((result + Number.EPSILON) * 1e10) / 1e10;
    return Number(rounded.toPrecision(12));
  } catch {
    return null;
  }
}

// Front-door: given a chat line, return a clean math answer or null.
export function trySolveMath(text) {
  const expr = extractMathExpression(text);
  if (!expr) return null;
  const value = evaluateMath(expr);
  if (value === null) return null;
  return `${expr} = ${value}`;
}

File: python/vierrataleai/utils/math.py
"""math.py - Deterministic, safe arithmetic resolver for VierrataleAI.

Parses arithmetic ( + - * / ^ % parentheses, decimals ) WITHOUT eval(), so it
can never execute arbitrary code. Returns None when the input is not a clean
math expression, so callers fall back to the model.
"""
import re

# Restrict to digits, decimal points, arithmetic operators and parens.
_ALLOWED = re.compile(r"^[0-9+\-*\/^%().\s]+$")


def extract_math_expression(text: str):
    """Pull a bare arithmetic expression out of a chat line, or None."""
    if not text:
        return None
    s = str(text).strip()
    # Strip a leading question/math verb phrase if present.
    s = re.sub(
        r"^(?:what(?:'s| is| are)|what's|compute|calculate|solve|evaluate|equals?)\b\s*",
        "",
        s,
        flags=re.IGNORECASE,
    ).strip()
    s = re.sub(r"[?,!.]$", "", s).strip()
    if not s:
        return None
    if not _ALLOWED.match(s):
        return None
    if not re.search(r"\d", s):
        return None
    if re.search(r"[a-zA-Z]", s):
        return None
    if not re.search(r"[\+\-\*\/\^%]", s):
        return None
    return s


def _tokenize(expr):
    tokens = []
    i = 0
    n = len(expr)
    while i < n:
        c = expr[i]
        if c.isspace():
            i += 1
            continue
        if c in "+-*/^%()":
            tokens.append(c)
            i += 1
            continue
        m = re.match(r"^(\d+(?:\.\d+)?|\.\d+)([eE][+-]?\d+)?", expr[i:])
        if not m:
            return None
        tokens.append(float(m.group(0)))
        i += len(m.group(0))
    return tokens


class _Parser:
    """Recursive-descent parser with correct precedence (right-assoc ^ power)."""

    def __init__(self, tokens):
        self.tokens = tokens
        self.pos = 0

    def peek(self):
        return self.tokens[self.pos] if self.pos < len(self.tokens) else None

    def next(self):
        t = self.peek()
        self.pos += 1
        return t

    def expect(self, val):
        t = self.next()
        if t != val:
            raise ValueError("syntax error")

    def parse_expr(self):
        return self.parse_term()

    def parse_term(self):
        left = self.parse_mul()
        while self.peek() in ("+", "-"):
            op = self.next()
            right = self.parse_mul()
            left = left + right if op == "+" else left - right
        return left

    def parse_mul(self):
        left = self.parse_power()
        while self.peek() in ("*", "/", "%"):
            op = self.next()
            right = self.parse_power()
            if op == "*":
                left = left * right
            elif op == "/":
                left = left / right
            else:
                left = left % right
        return left

    def parse_power(self):
        left = self.parse_unary()
        if self.peek() == "^":
            self.next()
            right = self.parse_power()
            left = left ** right
        return left

    def parse_unary(self):
        if self.peek() == "+":
            self.next()
            return self.parse_unary()
        if self.peek() == "-":
            self.next()
            return -self.parse_unary()
        return self.parse_factor()

    def parse_factor(self):
        t = self.peek()
        if t == "(":
            self.next()
            v = self.parse_expr()
            self.expect(")")
            return v
        if isinstance(t, float):
            self.next()
            return t
        raise ValueError("syntax error")


def evaluate_math(expr):
    try:
        tokens = _tokenize(expr)
        if not tokens:
            return None
        p = _Parser(tokens)
        result = p.parse_expr()
        if p.pos != len(tokens):
            return None
        if result != result or result in (float("inf"), float("-inf")):  # NaN / inf
            return None
        return round(result, 10)
    except Exception:
        return None


def try_solve_math(text):
    expr = extract_math_expression(text)
    if expr is None:
        return None
    try:
        value = evaluate_math(expr)
    except Exception:
        return None
    if value is None:
        return None
    return f"{expr} = {value}"
